### DISCUSSION TITLE
Welcome to Web-LLM-Assistant-Llamaccp-Ollama Discussions!

### DISCUSSION THREAD

**Discussion by TheBlewish** [TheBlewish/Web-LLM-Assistant-Llamacpp-Ollama#1] [Announcements]
Welcome to the Web-LLM-Assistant-Llama-cpp discussions!
I'm excited to share this project with you all. As a newcomer to open source development, I've created this assistant to combine the power of local LLMs with web search capabilities. The goal is to provide an easy-to-use, privacy-focused tool for enhanced information retrieval and analysis.
I'd love to hear your thoughts, ideas, and suggestions on how we can improve and expand this project. Your input, no matter what form it takes, is incredibly valuable. Whether you're a seasoned developer or just getting started, please feel free to contribute in any way you're comfortable with.
Don't hesitate to:

Share your ideas or suggestions
Report any issues you encounter
Offer feedback on the current functionality
Propose new features or directions for the project

Every bit of input helps shape the future of this tool. I'm eager to learn from your experiences and insights, so please don't hold back!
Thank you for taking the time to explore this project. Let's see where we can take it together!

**Comment by ali0une** [comment]
Hi there.

OK, i'll start then ...

First congrats for this application that works pretty good and is quite simple to install and use, love the use of GGUF.
i'm not a python coder myself, i can hack a bit but i'm very limited right now.

Things that could be improved :
CTRL+D to execute the prompt, i'd prefer just to hit Enter.
Some settings in a user conf file like n_ctx, max_tokens in Web-LLM.py and max_attempts in Self_Improving_Search.py

There is [ddg-web-search-chat space on HF](https://huggingface.co/spaces/poscye/ddg-web-search-chat) that basically does what your application does, i tried to get it to work locally and [failed](https://huggingface.co/spaces/poscye/ddg-web-search-chat/discussions/5) whatever i tried. There might be some black magic behind i did not understand ... anyway here is the [link to the code](https://huggingface.co/spaces/poscye/ddg-web-search-chat/tree/main) check the app.py file.

What would be really nice is to have a similar gradio UI for your application and the settings could be set directly in the webpage. Maybe you could adapt ddg-web-search-chat code to Web-LLM-Assistant-Llama.
A folder with the GGUF read by the application with a menu to select the model would be great. With the HF space it's in the Additional inputs drop down menu at the bottom.

**Reply by TheBlewish** [reply]
Hey thanks for your input! quick question when you say you tried to get it to work locally do you mean my program or the hugging face space?

Also the reason CTRL+D was used is because it allows you to hit enter to go down lines, which is the exact same as Llama.cpp's controls when your in multiline input mode, I could change it if you really want me to though! 

The idea to have a settings in a user config file is a good one, I will definitely be doing that, thanks for that suggestions, at the moment my number 1 priority is adding proper web-scraping to the program, so that instead of it being just very short snippets of results, the LLM will choose the best 2 results it sees and then the program will gather a bunch of information from those two links for the LLM to use to answer the User's query. 

What benefit is there to using a similar gradio UI for your application? The whole point is it runs locally and simply I don't know how to set it up like you suggested with that kind of UI!

**Comment by ali0une** [comment]
> do you mean my program or the hugging face space?

Both ; )

> Also the reason CTRL+D was used is because it allows you to hit enter to go down lines, which is the exact same as Llama.cpp's controls when your in multiline input mode

Oh ... didn't notice this, i mainly use llama.cpp through its web server.

i was really bothered not managing to make the ddg-web-search-chat space work locally, so i tried harder and i think i succeed. i even managed to make it work with my local GGUF (which was my main goal) so i think you can disregard my suggestion about Gradio and keep your application a cli.
[Gradio](https://www.gradio.app/docs) is a kind of web frontend for a python application.

> at the moment my number 1 priority is adding proper web-scraping to the program

You may like to have a look on how the ddg-web-search-chat space works, it uses [Beautiful soup and Selenium](https://huggingface.co/spaces/poscye/ddg-web-search-chat/blob/b2bc33826ce6854e078576a2e03563741a96fe0b/utils.py#L15)

**Reply by TheBlewish** [reply]
Thanks very much! for your suggestions! I have been hard at work and am going to be releasing a big update to the program likely later today or tomorrow at the latest! Web scraping has been successfully implemented already just been busy resolving bugs and adding more things to reduce the risk of bugs occurring with the LLM, I'm pretty excited this update should be pretty dam good compared to the current github state of the program, It's like well over double the amount of code and it makes it quite significantly better i'm hoping!

**Reply by ali0une** [reply]
Nice!

Also maybe you could try to enable GPU support it would make the application way faster.

**Reply by TheBlewish** [reply]
Yep we did that!
